demo_keyframe_bundle_adjustment_meta/apps/main_program/main_program.cpp

```diff
@@ -0,0 +1,268 @@
+  if (!lidar_data_file.is_open()) {
+    return false;
+  }
+  lidar_data_file.seekg(0, std::ios::end);
+  const size_t num_elements = lidar_data_file.tellg() / sizeof(float);
+  lidar_data_file.seekg(0, std::ios::beg);
+
+  std::vector<float> lidar_data_buffer(num_elements);
+  lidar_data_file.read(reinterpret_cast<char *>(&lidar_data_buffer[0]),
+                       num_elements * sizeof(float));
+
+  printf("Read %ld points in this lidar frame\n", lidar_data_buffer.size() / 4);
+
+  for (std::size_t i = 0; i < lidar_data_buffer.size(); i += 4) {
+    pcl::PointXYZI point;
+    point.x = lidar_data_buffer[i];
+    point.y = lidar_data_buffer[i + 1];
+    point.z = lidar_data_buffer[i + 2];
+    point.intensity = lidar_data_buffer[i + 3];
+    laser_cloud.push_back(point);
+
+    lidar_points.emplace_back(point.x, point.y, point.z);
+    lidar_intensities.push_back(point.intensity);
+  }
+  return true;
+}
+
+bool read_gt_pose(const std::string &value_string,
+                  const Eigen::Quaterniond &q_transform, float timestamp,
+                  nav_msgs::Odometry &odomGT, nav_msgs::Path &pathGT) {
+  std::stringstream pose_stream(value_string);
+  std::string s;
+  // Eigen::Matrix<double, 3, 4> gt_pose;
+  Eigen::Affine3d gt_pose;
+  geometry_msgs::PoseStamped poseGT;
+
+  for (std::size_t i = 0; i < 3; ++i) {
+    for (std::size_t j = 0; j < 4; ++j) {
+      std::getline(pose_stream, s, ' ');
+      if (s.empty()) return false;
+      gt_pose.matrix()(i, j) = stof(s);
+    }
+  }
+
+  Eigen::Quaterniond q_w_i(gt_pose.matrix().topLeftCorner<3, 3>());
+  Eigen::Quaterniond q = q_transform * q_w_i;
+  q.normalize();
+  Eigen::Vector3d t = q_transform * gt_pose.matrix().topRightCorner<3, 1>();
+
+  odomGT.header.stamp = ros::Time().fromSec(timestamp);
+  odomGT.pose.pose.orientation.x = q.x();
+  odomGT.pose.pose.orientation.y = q.y();
+  odomGT.pose.pose.orientation.z = q.z();
+  odomGT.pose.pose.orientation.w = q.w();
+  odomGT.pose.pose.position.x = t(0);
+  odomGT.pose.pose.position.y = t(1);
+  odomGT.pose.pose.position.z = t(2);
+
+  poseGT.header = odomGT.header;
+  poseGT.pose = odomGT.pose.pose;
+  pathGT.header.stamp = odomGT.header.stamp;
+  pathGT.poses.push_back(poseGT);
+  return true;
+}
+
+int main(int argc, char **argv) {
+  ros::init(argc, argv, "kitti_helper");
+  ros::NodeHandle publicNH;
+
+  ros::NodeHandle n("~");
+  std::string dataset_folder, sequence_number, output_bag_file;
+  n.getParam("dataset_folder", dataset_folder);
+  n.getParam("sequence_number", sequence_number);
+  std::cout << "Reading sequence " << sequence_number << " from "
+            << dataset_folder << '\n';
+  bool to_bag;
+  n.getParam("to_bag", to_bag);
+  if (to_bag) n.getParam("output_bag_file", output_bag_file);
+  int publish_delay;
+  n.getParam("publish_delay", publish_delay);
+
+  // Read the viso tracking configuration
+  std::string visoConfigFileName;
+  if (!n.getParam("viso_config", visoConfigFileName)) {
```

Comment:
We can read the config params for each module in this manner. That is, we can have the config yml file for each processing module and read the yml file and feed it into each module.